ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+            return model.generate(
+                *args,
+                **kwargs,
+                pad_token_id=model.config.pad_token_id,  # not always set properly by .generate()
+                # eos_token_id=model.config.eos_token_id,  TODO remove
+            )
+
+
+class TransformerHandler(ModuleHandler):
+    def __init__(self, config: TransformerHandlerConfig):
+        # Maybe this should be configurable? Hard to isolate the effect though
+        torch.set_float32_matmul_precision("high")
+
+        self.config = config
+        # Use local_rank to resolve model location only in the main process *for each node*
+        config.lm_config.resolve_model_location(is_main_process=self.local_rank == 0)
+
+        match config.lm_type:
+            case LMType.GENERATION:
+                tokenizer, model = config.lm_config.get_causal_lm()
+                # On left for https://github.com/huggingface/transformers/pull/7552
+                # ^ that seems to work for most HF models w/ absolute position embeddings
+                # Left padding always works for relative position embeddings
+                tokenizer.padding_side = "left"
+            case LMType.REGRESSION:
+                tokenizer, model = config.lm_config.get_regression_lm()
+            case _:
+                raise NotImplementedError(f"Unsupported LM type: {config.lm_type}")
+        tokenizer.pad_token = tokenizer.eos_token
+        super().__init__(model)
+        self.tokenizer = tokenizer
+        maybe_set_tokenizer_chat_template(self.tokenizer)
+
+        self._setup_accelerator()
+
+        if config.checkpoint is not None:
+            self.load_checkpoint(config.checkpoint)
+
+    def _setup_accelerator(self):
+        self.accelerator = accelerate.Accelerator(
+            # This has to be disabled because accelerator wraps forward() to upcast outputs to fp32. That
+            # causes problems with generation, where the cache is expected to be in the same dtype as the model.
+            # TODO: understand why this doesn't break in huggingface code and re-enable in ours.
+            # mixed_precision=("bf16" if self.config.lm_config.dtype == TorchDType.bf16 else "no")
+        )
+        self.module = self.accelerator.prepare(self.module)
+
+    @property
+    def local_rank(self) -> int:
+        return int(os.getenv("LOCAL_RANK", "0"))
+
+    def load_checkpoint(self, ckpt: os.PathLike | str, **kwargs) -> None:
+        logger.info(f'Loading checkpoint from "{ckpt}"')
+        self.accelerator.load_state(str(ckpt), **kwargs)
+
+    def save_checkpoint(self, ckpt: os.PathLike | str, **kwargs) -> None:
+        self.accelerator.save_state(str(ckpt), **kwargs)
+        self.barrier()
+        # We do not want to save random states - they would be loaded by load_state
+        # automatically. Clean up after all processes have saved.
+        if int(os.getenv("RANK", "0")) == 0:
+            random_states = Path(ckpt).rglob("random_state*pkl")
```

Comment:
```suggestion
            random_states = Path(ckpt).rglob("random_state*.pkl")
```

Just capturing extension